File: web/src/lib/sentry.ts
/**
 * Frontend Sentry init stub (ISSUE-078).
 *
 * Architecture §12 calls for Sentry on errors. We keep the init thin
 * and DSN-gated: when ``NEXT_PUBLIC_SENTRY_DSN`` is unset, ``initSentry``
 * is a no-op so previews / local dev never ship telemetry.
 *
 * The actual ``@sentry/nextjs`` package isn't bundled at this point in
 * the project — wiring lives in this module so that when the real SDK
 * lands (with build-time sourcemap upload via ``next.config.mjs``), only
 * the body of ``initSentry`` changes; call sites stay stable.
 *
 * PRD-Ref: NFR-016.
 */

export interface SentryInitOptions {
  dsn?: string | null;
  environment?: string;
  release?: string;
  /**
   * Optional ``beforeSend`` hook that mutates the event before transport.
   * Defaults to {@link scrubEvent} which strips known PII patterns from
   * any event property (birth_dt, payment keys, Toss order IDs, JWTs).
   */
  beforeSend?: (event: SentryEventLike) => SentryEventLike | null;
}

export interface SentryEventLike {
  message?: string;
  extra?: Record<string, unknown>;
  request?: { headers?: Record<string, unknown> };
  breadcrumbs?: { values?: Array<Record<string, unknown>> };
  [key: string]: unknown;
}

export const REDACTED = '[REDACTED]';

const SENSITIVE_KEYS = new Set([
  'birth_dt',
  'birth_date',
  'birthdate',
  'birth',
  'payment_key',
  'paymentkey',
  'card_number',
  'cardnumber',
  'card_no',
  'cvv',
  'secret_key',
  'secretkey',
  'api_key',
  'apikey',
  'access_token',
  'refresh_token',
  'jwt',
  'authorization',
  'password',
  'passwd',
]);

const STRING_PATTERNS: Array<[RegExp, string]> = [
  // JWT — 3 base64url segments
  [/\beyJ[A-Za-z0-9_-]+={0,2}\.[A-Za-z0-9_-]+={0,2}\.[A-Za-z0-9_-]+={0,2}\b/g, REDACTED],
  // Toss order id (ORD-...)
  [/\b(?:ORD[-_])[A-Za-z0-9_-]{6,}\b/g, REDACTED],
  // paymentKey value (long alnum after `paymentKey":"` / `paymentKey=`)
  [/(paymentKey["':= ]+)[A-Za-z0-9_-]{16,}/gi, `$1${REDACTED}`],
  // birth_dt= / "birth_dt": ISO date
  [
    /(birth[_-]?(?:dt|date)\s*[=:]\s*["']?)\d{4}-\d{2}-\d{2}(?:[T ]\d{2}:\d{2}(?::\d{2})?)?/gi,
    `$1${REDACTED}`,
  ],
];

function scrubValue(value: unknown): unknown {
  if (value === null || value === undefined) return value;
  if (typeof value === 'string') {
    let out = value;
    for (const [pattern, replacement] of STRING_PATTERNS) {
      out = out.replace(pattern, replacement);
    }
    return out;
  }
  if (typeof value === 'number' || typeof value === 'boolean') return value;
  if (Array.isArray(value)) return value.map(scrubValue);
  if (typeof value === 'object') {
    const out: Record<string, unknown> = {};
    for (const [k, v] of Object.entries(value as Record<string, unknown>)) {
      out[k] = SENSITIVE_KEYS.has(k.toLowerCase()) ? REDACTED : scrubValue(v);
    }
    return out;
  }
  return value;
}

/**
 * ``before_send`` callback — recursively scrubs PII from the event tree.
 * Returns the (possibly mutated) event so transport proceeds; never
 * returns null (we always want error visibility).
 */
export function scrubEvent(event: SentryEventLike): SentryEventLike {
  return scrubValue(event) as SentryEventLike;
}

let _initialised = false;

/**
 * Initialise Sentry. Returns ``true`` iff init actually happened.
 *
 * Idempotent — calling twice is safe. When ``dsn`` is unset/empty the
 * SDK is never touched so test runs stay hermetic.
 */
export function initSentry(options: SentryInitOptions): boolean {
  const dsn = options.dsn?.trim();
  if (!dsn) return false;
  if (_initialised) return true;

  // The real ``@sentry/nextjs`` import lands here once the dependency
  // is added at deploy time. Until then we record the init parameters
  // so consumers (and tests) can verify the wiring without bundling the
  // SDK into the dev build.
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  (globalThis as any).__voicesaju_sentry_init__ = {
    dsn,
    environment: options.environment ?? 'local',
    release: options.release ?? null,
    beforeSend: options.beforeSend ?? scrubEvent,
  };
  _initialised = true;
  return true;
}

/**
 * Test-only reset. Avoid using in production code.
 */
export function _resetSentryForTests(): void {
  _initialised = false;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  delete (globalThis as any).__voicesaju_sentry_init__;
}

File: api/voicesaju/observability/sentry.py
"""Sentry SDK init + PII scrubbing (ISSUE-078).

Architecture §12 calls for Sentry on errors. Architecture §11 + NFR-005
forbid PII (birth dates, payment keys, Toss order IDs) from leaving the
backend in any telemetry stream — including error reports.

Init is gated by ``Settings.sentry_dsn``: when unset the SDK is never
initialised, so ``import voicesaju.observability.sentry`` has zero
runtime impact in tests and local dev.

PRD-Ref: NFR-016 (uptime / error visibility).
"""

from __future__ import annotations

import re
from typing import Any

import sentry_sdk
from sentry_sdk.integrations.fastapi import FastApiIntegration
from sentry_sdk.integrations.starlette import StarletteIntegration

# Keys that must never leave the backend in a Sentry event. Lowercased
# during comparison so ``Birth_DT`` and ``paymentKey`` are caught too.
_SENSITIVE_KEYS: frozenset[str] = frozenset(
    {
        "birth_dt",
        "birth_date",
        "birthdate",
        "birth",
        "payment_key",
        "paymentkey",
        "card_number",
        "cardnumber",
        "card_no",
        "cvv",
        "secret_key",
        "secretkey",
        "api_key",
        "apikey",
        "access_token",
        "refresh_token",
        "jwt",
        "authorization",
        "password",
        "passwd",
    }
)

REDACTED = "[REDACTED]"

# String-shape patterns we never want to ship to Sentry. We re-use the
# same families as the logging redactor (ISSUE-079) so the two filters
# stay consistent; we don't import from logging.py to keep this module
# usable even before logging is configured.
_STRING_PATTERNS: tuple[tuple[re.Pattern[str], str], ...] = (
    (
        re.compile(
            r"\beyJ[A-Za-z0-9_-]+={0,2}\.[A-Za-z0-9_-]+={0,2}\.[A-Za-z0-9_-]+={0,2}\b"
        ),
        REDACTED,
    ),
    (re.compile(r"\b(?:ORD[-_])[A-Za-z0-9_-]{6,}\b"), REDACTED),
    (
        re.compile(
            r"(paymentKey[\"':= ]+)[A-Za-z0-9_-]{16,}",
            re.IGNORECASE,
        ),
        rf"\1{REDACTED}",
    ),
    (
        re.compile(
            r"(birth[_-]?(?:dt|date)\s*[=:]\s*[\"']?)"
            r"\d{4}-\d{2}-\d{2}(?:[T ]\d{2}:\d{2}(?::\d{2})?)?",
            re.IGNORECASE,
        ),
        rf"\1{REDACTED}",
    ),
)


def _scrub_value(value: Any) -> Any:
    """Recursively scrub PII from any JSON-like value."""
    if value is None or isinstance(value, (bool, int, float, bytes)):
        return value
    if isinstance(value, str):
        out = value
        for pattern, replacement in _STRING_PATTERNS:
            out = pattern.sub(replacement, out)
        return out
    if isinstance(value, dict):
        return {
            k: (
                REDACTED
                if isinstance(k, str) and k.lower() in _SENSITIVE_KEYS
                else _scrub_value(v)
            )
            for k, v in value.items()
        }
    if isinstance(value, (list, tuple)):
        cleaned = [_scrub_value(item) for item in value]
        return type(value)(cleaned) if isinstance(value, tuple) else cleaned
    return value


def scrub_event(
    event: dict[str, Any], _hint: dict[str, Any] | None = None
) -> dict[str, Any]:
    """Sentry ``before_send`` callback — scrub PII from the event tree.

    Returns the (possibly mutated) event dict so Sentry transmits the
    redacted copy. We never return ``None`` (which would drop the event)
    — visibility of an error is more valuable than fully suppressing it.
    """
    return _scrub_value(event)


def init_sentry(
    *,
    dsn: str | None,
    environment: str = "local",
    release: str | None = None,
    traces_sample_rate: float = 0.0,
) -> bool:
    """Initialise the Sentry SDK if ``dsn`` is set.

    Returns ``True`` iff initialisation actually happened. The factory
    in :func:`voicesaju.main.create_app` uses the return value only for
    logging; absence of a DSN is the normal path in local/test runs.

    The init pins ``send_default_pii=False`` and registers
    :func:`scrub_event` as the ``before_send`` hook so even if a user-
    supplied PII field slips past the regular logging filter, it is
    redacted before transmission.
    """
    if not dsn:
        return False

    sentry_sdk.init(
        dsn=dsn,
        environment=environment,
        release=release,
        send_default_pii=False,
        # Keep traces opt-in — the OTel pipeline (ISSUE-077) is the
        # primary trace backend; Sentry traces are a backup only.
        traces_sample_rate=traces_sample_rate,
        integrations=[
            FastApiIntegration(transaction_style="endpoint"),
            StarletteIntegration(transaction_style="endpoint"),
        ],
        before_send=scrub_event,
    )
    return True


__all__ = [
    "REDACTED",
    "init_sentry",
    "scrub_event",
]
